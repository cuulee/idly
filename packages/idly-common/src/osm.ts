export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
