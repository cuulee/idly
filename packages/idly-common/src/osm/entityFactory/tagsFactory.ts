import { sortObjectKeys } from '../../misc/sortObjectKeys';
import { Tags } from '../structures';

// @TOFIX make tags frozen objects.
// and also figure out how to make reserved keys work in object
// or use so other data structure.
const emptyTag = {};
export function tagsFactory(
  rawTags: { [key: string]: string } = emptyTag
): Tags {
  return sortObjectKeys(rawTags);
}
