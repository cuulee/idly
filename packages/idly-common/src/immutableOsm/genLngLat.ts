import { recordify } from "typed-immutable-record";
import { $LngLat } from "../immutableOsm/immutableOsm";
import { LngLat } from "../osm/structures";

export function $genLngLat(l: LngLat): $LngLat {
  return recordify<LngLat, $LngLat>(l);
}
