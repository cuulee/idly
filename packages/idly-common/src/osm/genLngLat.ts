import { LngLat } from "../osm/structures";

export function genLngLat(
  obj: { lon: number; lat: number } | [number, number],
): LngLat {
  if (Array.isArray(obj)) {
    return { lon: obj[0], lat: obj[1] };
  }
  return { ...obj };
}
