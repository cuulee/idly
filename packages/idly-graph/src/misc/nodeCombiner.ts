import { Feature } from 'idly-common/lib/osm/feature';
import { Node } from 'idly-common/lib/osm/structures';

export function nodeCombiner(node: Node, existingProps: {}): Feature<any, any> {
  return {
    id: node.id,
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [node.loc.lon, node.loc.lat],
    },
    properties: {
      ...existingProps,
      id: node.id,
    },
  };
}
