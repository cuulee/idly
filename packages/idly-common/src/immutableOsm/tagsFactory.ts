import { Tags } from '../osm/structures';
import { $Tags } from '../immutableOsm/immutableOsm';
import { Map as $Map } from 'immutable';

export function $tagsFactory(t: Tags): $Tags {
  return $Map(t);
}
