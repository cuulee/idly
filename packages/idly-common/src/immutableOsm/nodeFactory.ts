import { $attributesGen } from './attributesGen';
import { $genLngLat } from './genLngLat';
import { $tagsFactory } from './tagsFactory';
import { $Node, $__Node } from '../immutableOsm';
import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import { Attributes, EntityId, EntityType, LngLat, Node, Tags } from '../osm';
import { recordify } from 'typed-immutable-record';

export function nodeFactory(n: Node): $Node {
  var $tags = $tagsFactory(n.tags);
  var $loc = $genLngLat(n.loc);
  var $attr = $attributesGen(n.attributes);
  return recordify<$__Node, $Node>({
    id: n.id,
    tags: $tags,
    type: EntityType.NODE,
    loc: $loc,
    attributes: $attr
  });
}
