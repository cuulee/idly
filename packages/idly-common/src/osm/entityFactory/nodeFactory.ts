import { lngLatFactory } from '../../geo/lngLatFactory';
import { deepFreeze } from '../../misc/deepFreeze';

import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function nodeFactory(
  {
    id,
    tags = tagsFactory(),
    loc = lngLatFactory([0, 0]),
    attributes = attributesFactory(),
  }: {
    id: EntityId;
    tags?: Tags;
    loc?: LngLat;
    attributes?: Attributes;
  },
  freeze = true
): Node {
  return deepFreeze<Node>(
    {
      attributes: attributesFactory(attributes),
      id,
      loc,
      tags: tagsFactory(tags),
      type: EntityType.NODE,
    },
    freeze
  );
}
