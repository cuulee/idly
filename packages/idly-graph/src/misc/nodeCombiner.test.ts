import { genLngLat } from 'idly-common/lib/osm/genLngLat';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { tagsFactory } from 'idly-common/lib/osm/tagsFactory';
import { nodeCombiner } from './nodeCombiner';

const n1 = nodeFactory({ id: 'n-1' });
const n11 = nodeFactory({
  id: 'n-1',
  loc: genLngLat({ lon: 15, lat: 10 }),
  tags: tagsFactory([['k', 'k']]),
});

describe('converts node to feat', () => {
  describe('nodeCombiner', () => {
    it('should behave...', () => {
      expect(nodeCombiner(n1, {})).toMatchSnapshot();
      expect(nodeCombiner(n11, {})).toMatchSnapshot();
    });
    it('should add computed Props', () => {
      expect(nodeCombiner(n1, { test: '1234', h: 3 })).toMatchSnapshot();
    });
  });
});
