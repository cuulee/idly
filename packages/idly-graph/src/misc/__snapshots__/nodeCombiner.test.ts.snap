// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts node to feat nodeCombiner should add computed Props 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "h": 3,
    "id": "n-1",
    "test": "1234",
  },
  "type": "Feature",
}
`;

exports[`converts node to feat nodeCombiner should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "id": "n-1",
  },
  "type": "Feature",
}
`;

exports[`converts node to feat nodeCombiner should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "id": "n-1",
  },
  "type": "Feature",
}
`;
